### ISSUE TITLE
Valid Polygon with Hole renders with clipping Artifacts

### ISSUE THREAD

**Issue by KiwiKilian** [maplibre/maplibre-gl-js#4367]
**maplibre-gl-js version**: Reproduced with 4.5.0 as well as 1.15.3

**browser**: Chromium 126.0.6478.127

### Steps to Trigger Behavior

We have a polygon with a hole to create a mask. Basically it's a large rectangle with a more detailed polygon within. On some Zoom levels, clipping artifacts appear. The example is a very much simplified polygon. The simplified version was validated with https://geojsonlint.com/ – therefore winding order shouldn't be a problem.

### Link to Demonstration

https://codepen.io/KiwiKilian/pen/QWXLVaG?editors=1111

### Expected Behavior

Clipping artifacts should not appear.

### Actual Behavior

Filled artifact areas appear and disappear.

https://github.com/maplibre/maplibre-gl-js/assets/18724559/205bce4d-e6b7-4388-a51a-85b38ed32fc1

**Comment by HarelM** [comment]
Isn't this a duplicate of #4357 ?

**Comment by KiwiKilian** [comment]
No, that's why I've emphasized **valid** polygon in the title. #4357 is about an invalid polygon, only three positions, first and last don't match. But the outcome looks a bit similar.

**Comment by HarelM** [comment]
Does this problem exists for vector tiles source as well?

**Comment by KiwiKilian** [comment]
Do you have a recommendation for the easiest way to test this with a vector source?

**Comment by HarelM** [comment]
Aren't the demo tiles such a source?
https://github.com/maplibre/demotiles

**Comment by wipfli** [comment]
You could feed the geojson from this issue to tippecanoe and create a small mbtiles containing mvts. See https://github.com/felt/tippecanoe

**Comment by KiwiKilian** [comment]
I went with `pmtiles` because they can easily be served with `npx serve` – I hope this as also fine. I used no options which would change the polygon.

The problem does not occur with a `vector` source like `pmtiles`. I've overlayed a red `geojson` source with a green `vector` source, both with 50 % opacity. This is the result:

https://github.com/maplibre/maplibre-gl-js/assets/18724559/8cc9f39b-1a5b-48bf-9cc6-a4d578bdc93e

To clearly show, this is the `vector` source by itself:

https://github.com/maplibre/maplibre-gl-js/assets/18724559/471b076b-97a4-4fe7-96a5-a797ec20ba93

You can [check out the reproduction here](https://github.com/KiwiKilian/maplibre-gl-js-polygon-hole-bug).

**Comment by HarelM** [comment]
So if I needed to guess the problem is with the geojson to vector tile conversion happening in geojson-vt package, which is a dependency of this library. 
I would recommend to try and reproduce this with only geojson-vt package to create a minimal reproduction, after that we'll need to see if we can fix it upstream or we might need to fork.

**Comment by KiwiKilian** [comment]
I tried their [debug page](https://mapbox.github.io/geojson-vt/debug/) with the [polygon](https://github.com/KiwiKilian/maplibre-gl-js-polygon-hole-bug/blob/main/polygon.geojson), but it looked fine. It definitely shows the problematic area/coordinates span across multiple tiles. Would you expect to reproduce on that page?

**Comment by PhungVanHoa** [comment]
The problem does not occur when change maxzoom lower than 11 

https://github.com/maplibre/maplibre-gl-js/assets/16158238/9e7ca3af-1461-490a-a010-4e3a51aa0be9
